pyHalo/realization_extensions.py

```diff
@@ -328,4 +391,103 @@ def add_primordial_black_holes(self, pbh_mass_fraction, kwargs_pbh_mass_function
                                                         r_max_arcsec, arcsec_per_pixel)
 
         return realization_with_clustering.join(realization_smooth)
+      
+def _get_number_flucs(realization,de_Broglie_wavelength,shape,args):
+    """
+    This function returns the number of fluctuations to place in the realization.
+
+    :param realization: the realization to which to add the fluctuations
+    :param de_Broglie_wavelength: de Broglie wavelength of the DM particle in kpc
+    :param shape: keyword argument for fluctuation geometry, see 'add_ULDM_fluctuations'
+    :param args: properties of the given shape, see 'add_ULDM_fluctuations'
+    """
+
+    D_d = realization.lens_cosmo.cosmo.D_A_z(realization._zlens)
+    arcsec = realization.lens_cosmo.cosmo.arcsec
+    fluc_area=np.pi*de_Broglie_wavelength**2 #de Broglie area
+    to_kpc = D_d * arcsec * 1e3 # convert arcsec to kpc
+
+    if shape=='ring': # fluctuations in a circular slice (for visualization purposes)
+
+        rmin_kpc,rmax_kpc = args['rmin'] * to_kpc, args['rmax'] * to_kpc #args in kpc
+        area_ring = np.pi*(rmax_kpc**2-rmin_kpc**2) # volume of ring
+        n_flucs_expected=area_ring/fluc_area # number of fluctuations in ring
+        n_flucs = np.random.poisson(n_flucs_expected)
+    
+    if shape=='ellipse': # fluctuations in a elliptical slice (for visualization purposes)
+        
+        amin_kpc,bmin_kpc,amax_kpc,bmax_kpc=args['amin'] * to_kpc, args['bmin'] * to_kpc, args['amax'] * to_kpc, args['bmax'] * to_kpc #args in kpc
+        area_ellipse=np.pi*(amax_kpc*bmax_kpc - amin_kpc*bmin_kpc) # volume of ellipse
+        n_flucs_expected=area_ellipse/fluc_area # number of fluctuations in ellipse
+        n_flucs = np.random.poisson(n_flucs_expected)
+
+    if shape=='aperture': # fluctuations around lensing images (for computation)
+
+        n_images=len(args['x_images']) #number of lensed images
+        r_kpc = args['aperture'] * D_d * arcsec * 1e3 #aperture in kpc
+        area_aperture = np.pi*r_kpc**2 # aperture area
+        n_flucs_expected = area_aperture/fluc_area #number of expected fluctuations per aperture
+        n_flucs = np.random.poisson(n_flucs_expected,n_images) #draw number of fluctuations from poisson distribution for each image
+        n_flucs = n_flucs[n_flucs!=0] #get rid of aperture if zero fluctuations within it
+
+    return n_flucs
+
+def _get_fluctuation_halos(realization,amplitude_std_dev,gaussian_std_dev,shape,n_flucs,args):
```

Comment:
can you define the amplitude_std_dev already in units of convergence? then you don't have to make calls to LensCosmo to compute sigma crit and angular diameter distances inside this function. Instead, you can make do computations inside the preset model class, and then pass in the fluctuation amplitudes in units of convergence